### ISSUE TITLE
ADR 2: How to show test results in the html report

### ISSUE THREAD

**Issue by camilamaia** [scanapi/scanapi#161]
## Architecture Decision Review - ADR

- How are we going to show the tests in the html report
- Which color should we use?
- How are we going to show each test case?
- How are we going to show if a test passed?
- How are we going to show if a test failed?
- Should we use HTTP methods with color? Yes or No?

This discussion started [here](https://github.com/scanapi/scanapi/pull/157#pullrequestreview-424762095)

Related ADR: #164

**Comment by camilamaia** [comment]
I did some tests, they are all ugly 😂  But maybe we can get some ideas of what works and what doesn't work:

https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version1.html
https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version2.html
https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version3.html
https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version4.html
https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version5.html
https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version6.html

I can't attach HTML files, that is why a created a temporary branch just to load this tests.

@barbosa @djalmaaraujo @gillianomenezes any thought?

**Comment by camilamaia** [comment]
Basically we have three different test status we need to show

1. The overview. If **all** tests passed. Or, if **any** test failed

![image](https://user-images.githubusercontent.com/2728804/85133428-0ec1ee00-b211-11ea-8a1c-e1ebc1bdf166.png)

If all tests passed, the test summary should, somehow, show that everything is ok and there is no need to check anything else. It is all green.

2. The Endpoint. We need to show if **all** tests for that specific endpoint passed (or, if **any** failed)

![image](https://user-images.githubusercontent.com/2728804/85133624-6d876780-b211-11ea-81f4-45e1b88b3c6c.png)

Here it is important to provide somehow an easy and fast way for the user to find which endpoint has a failing test. Let's say we have a report with dozens of endpoints. We don't want the user opening all the endpoints trying to find which one failed.

Something written would be a solution. Let's say there is something like this

![image](https://user-images.githubusercontent.com/2728804/85133896-f2728100-b211-11ea-84da-8dd4ed345154.png)

The user can easily use `ctrl + f` and search for `[failed]` to find the endpoints with failing tests.

3. The test itself. 

![image](https://user-images.githubusercontent.com/2728804/85134201-7d537b80-b212-11ea-91c5-61e18a4fded2.png)

---

We are using different colors for each HTTP method. We want to make clear that they are different. 

![image](https://user-images.githubusercontent.com/2728804/85134572-1aaeaf80-b213-11ea-88d2-f94b351c9f3a.png)

But, as @barbosa mentioned [here](https://github.com/scanapi/scanapi/pull/157#pullrequestreview-424762095), "green/yellow/red colors may give the impression of success/failure for each of the requests being executed."

So how we can show this difference without making this impression of success/failure/skipped? Which colors should we use? I tried some pastel colors, but I did not like the ones I picked

![image](https://user-images.githubusercontent.com/2728804/85134692-4e89d500-b213-11ea-9814-c88143be04f5.png)

Current we only support [five HTTP methods](https://github.com/scanapi/scanapi/blob/master/scanapi/tree/request_node.py#L34), but there are [9 in total](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).

---

In summary, it would be nice to discuss:

1. What is the best way to show the overview status?
2, What is the best way to show the endpoint status?
3. What is the best way to show the test status?
4. What is the best way to differentiate the HTTP methods?

**Comment by djalmaaraujo** [comment]
@camilamaia @barbosa 
Hey, that's a good discussion. Here's what I would prefer to see in this case:

- I would suggest red/green for the main status of the runner (PASSED/FAILED)
- One-color only for the HTTP methods. I am not sure if the different colors make sense for the report, to be honest. So maybe we could use BLUE (maybe) since blue is not a TRAFFIC color (red/green/yellow)
- For HTTP status 200/404/etc, I would follow the standard industry: 200/2** goes GREEN, anything bad 400/500, etc goes red.

I can provide the palette for this if you agree.

**Comment by camilamaia** [comment]
Versions with images: 

## Version 1
![version1_closed](https://user-images.githubusercontent.com/2728804/85138260-dfaf7a80-b218-11ea-8406-ab67b962b6af.png)
![version1_opened](https://user-images.githubusercontent.com/2728804/85138265-e0e0a780-b218-11ea-8e8a-7b530d1e5a1c.png)

## Version 2
![version2_closed](https://user-images.githubusercontent.com/2728804/85138267-e1793e00-b218-11ea-8f24-19227da32610.png)
![version2_opened](https://user-images.githubusercontent.com/2728804/85138269-e211d480-b218-11ea-8c35-fc9c8954723a.png)

## Version 3
![version3_closed](https://user-images.githubusercontent.com/2728804/85138273-e2aa6b00-b218-11ea-91b7-207b74ec4fd0.png)
![version3_opened](https://user-images.githubusercontent.com/2728804/85138277-e3430180-b218-11ea-9e6f-680bf0aa6926.png)

## Version 4
![version4_closed](https://user-images.githubusercontent.com/2728804/85138279-e3db9800-b218-11ea-96a1-4818df065b7f.png)
![version4_opened](https://user-images.githubusercontent.com/2728804/85138283-e4742e80-b218-11ea-96d5-2b8122e9fd79.png)

## Version 5
![version5_closed](https://user-images.githubusercontent.com/2728804/85138287-e50cc500-b218-11ea-873c-e7772295d83d.png)
![version5_opened](https://user-images.githubusercontent.com/2728804/85138288-e5a55b80-b218-11ea-8ce7-1019b966442f.png)

## Version 6
![version6_closed](https://user-images.githubusercontent.com/2728804/85138291-e63df200-b218-11ea-98a2-8f10c8261f3d.png)
![version6_opened](https://user-images.githubusercontent.com/2728804/85138292-e6d68880-b218-11ea-8fd5-2a17a4f78048.png)

**Comment by gillianomenezes** [comment]
> @camilamaia @barbosa
> Hey, that's a good discussion. Here's what I would prefer to see in this case:
> 
> * I would suggest red/green for the main status of the runner (PASSED/FAILED)
> * One-color only for the HTTP methods. I am not sure if the different colors make sense for the report, to be honest. So maybe we could use BLUE (maybe) since blue is not a TRAFFIC color (red/green/yellow)
> * For HTTP status 200/404/etc, I would follow the standard industry: 200/2** goes GREEN, anything bad 400/500, etc goes red.
> 
> I can provide the palette for this if you agree.

I agree with this approach. IMHO the HTTP verbs in black bold seem better. This is why I vote in the  [first option](https://github.com/scanapi/scanapi/blob/adr2/testing_reports/version1.html), @camilamaia.

**Comment by barbosa** [comment]
> @camilamaia @barbosa
> Hey, that's a good discussion. Here's what I would prefer to see in this case:
> 
> * I would suggest red/green for the main status of the runner (PASSED/FAILED)
> * One-color only for the HTTP methods. I am not sure if the different colors make sense for the report, to be honest. So maybe we could use BLUE (maybe) since blue is not a TRAFFIC color (red/green/yellow)
> * For HTTP status 200/404/etc, I would follow the standard industry: 200/2** goes GREEN, anything bad 400/500, etc goes red.
> 
> I can provide the palette for this if you agree.

Yeah, I agree with this except for the last item. I believe color variation should be used for one thing and one thing only: indicate success or failure of test results.

I would keep both HTTP methods and status codes in black.

Mixing red/green in more than one context would cause confusion, I think.

**Comment by camilamaia** [comment]
@djalmaaraujo @gillianomenezes and @barbosa, thank you very much for your feedback and opinions. Based on then, I created a new version. What do you think?

### Failed Closed
![failed-closed](https://user-images.githubusercontent.com/2728804/85232338-9441cb80-b3d4-11ea-85cf-b83462747793.png)

### Failed Opened
![failed-opened](https://user-images.githubusercontent.com/2728804/85232343-97d55280-b3d4-11ea-9b45-a2b5b3a60b13.png)

### Passed Closed
![passed-closed](https://user-images.githubusercontent.com/2728804/85232344-986de900-b3d4-11ea-9cd8-3eec469982d7.png)

### Passed Opened
![passed-opened](https://user-images.githubusercontent.com/2728804/85232345-986de900-b3d4-11ea-859d-73e854543bdc.png)

**Comment by camilamaia** [comment]
@djalmaaraujo give me some help! Look how it is now:

![passed](https://user-images.githubusercontent.com/2728804/85330168-3c759400-b4aa-11ea-9d39-683f42d41896.png)
![error-opened](https://user-images.githubusercontent.com/2728804/85330072-13ed9a00-b4aa-11ea-97d9-9865e47ce959.png)

**Comment by camilamaia** [comment]
@gillianomenezes and @barbosa can I close this ticket? Do you have any other opinions? Thanks!

**Comment by barbosa** [comment]
<img width="184" alt="Screen Shot 2020-06-24 at 8 39 44 PM" src="https://user-images.githubusercontent.com/235208/85640841-e1cf6a00-b65a-11ea-937f-72b2860076c2.png">

it seems the icons are not vertically centered but we can fix this later

**Comment by camilamaia** [comment]
> <img alt="Screen Shot 2020-06-24 at 8 39 44 PM" width="184" src="https://user-images.githubusercontent.com/235208/85640841-e1cf6a00-b65a-11ea-937f-72b2860076c2.png">
> 
> it seems the icons are not vertically centered but we can fix this later

Great catch 🔍 👀

**Comment by camilamaia** [comment]
@GitVitor check this issue here! It might help you with the new changes on the report 🎨

**Comment by GitVitor** [comment]
@camilamaia thanks, i had never seen this ADR before;

Just to document, i had some thoughts about how we could improve our UI;

Below a list with some changes that i have work:

1. Change P(assed) or F(ailed) icon;
2. Change font-family to use [Gotham](https://www.cdnfonts.com/gotham.font);
3. Use diferent colors to represent each HTTP Method (Image 1);
4. Make request content, request response and request test visually grouped (Image 2);
5. Change Green to use the following hex `#48c78e`;
6. Change Red to use the following hex `#f14668`;


Image 1
![image](https://user-images.githubusercontent.com/15002664/130307511-8fd40bfe-b95f-4102-8173-cba2c97699e0.png)

Image 2
![image](https://user-images.githubusercontent.com/15002664/130307722-5d8c9f29-ad62-48a0-b45f-4b1e19052b96.png)

**Comment by GitVitor** [comment]
Based on comments that was sent here, maybe we could apply following changes:

1. Change P(assed) or F(ailed) icon;
2. Change font-family to use Gotham;
~Use diferent colors to represent each HTTP Method;~
~Make request content, request response and request test visually grouped;~
3. Change Green to use the following hex #48c78e;
4. Change Red to use the following hex #f14668;
